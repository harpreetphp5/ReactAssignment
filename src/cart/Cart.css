.Cart {
    width: 70%;
    margin: 100px auto;
}
ul li{
    list-style: none;
}
ul.cartHeader{
    display: flex;
    justify-content: space-between;
}

li.productTab{
    flex: 0.6;
    text-align: left;
}
li.qtyTab {
    flex: 0.2;
    text-align: left;
    flex-direction: row;
    margin-top: 10px;
}
li.totalTab, li.actionTab {
    flex: 0.1;
    text-align: left;
    margin-top: 10px;
    justify-content: center;
    align-items: center;
}
ul.cartHeader li {
    font-size: 12px;
    color: #aaa;
}
.cartPage, .cartWrap{
    display: flex;
    flex-direction: row;

}
.cartPage img{
    width: 70px;
    height: 70px;
}

input.qtyType {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 25px;
    background: #aaa;
    border: none;
    padding: 15px;
}
a.continueShopping {
    font-size: 12px;
    padding-bottom: 2px;
    border-bottom: 1px solid #333;
}
div.qtyBox{
    float: left;
}
div.qtyBtns {
    width: 23px;
    height: 18px;
    display: flex;
    flex-direction: column;

}
div.qtyBtns button.second {
    margin-top: 2px
}
.priceTotal, .priceSubtotal {
    display: flex;
    justify-content: space-between;
}
.cartOverview{
    width: 40%;
    float:right;
    margin-bottom: 20px;
}
.cartOverview li {
    margin-top: 10px;
    font-size: 12px;
    color: #aaa;
}
.cartOverview li.qtyTab {
    margin-top: 10px;
    margin-bottom: 10px;
    height: 15px;
}
.clear {
    clear: both;
}
.crossDelSimple {
    border: none;
    background: #fff;
    font-size: 19px;
    color: #aaa;
}
