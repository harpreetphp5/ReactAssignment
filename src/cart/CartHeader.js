import React from 'react';

const CartHeader = () => {
    return(
        <>
            <li className = "productTab" > PRODUCT</li>
            <li className="qtyTab">QUANTITY</li>
            <li className="totalTab">TOTAL</li>
            <li className="actionTab">ACTION</li>
        </>
    )
}

export default CartHeader;