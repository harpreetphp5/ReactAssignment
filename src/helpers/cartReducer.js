import React from 'react';

export const reducer = (state, action) => {
    switch (action.type) {
        case 'ADD_CART':
            return [...state, action.payload];
        case 'UPDATE_CART':
            return state.map(p => (p.title == action.payload.title) ? ({ ...p, qty: p.qty + action.payload.qty }) : p)
        case 'DELETE_CART':
            return state.filter(e => e.title != action.payload);
        default:
            return state;
    }
}