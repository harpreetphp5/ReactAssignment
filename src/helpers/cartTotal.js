import React from 'react';
// returns total sum of price of the all items in the cart 
const getTotalPrice = (cartStuff) => {
    var total = 0;
    cartStuff.map(stuff => (
        total += (stuff.qty*stuff.items.price))
    )
    return total;
}
// returns total number of items in the cart
const getTotalQty = (cartStuff) => {
    var total = 0;
    cartStuff.map(stuff => (
        total += stuff.qty)
    )
    return total;
}

export {getTotalPrice, getTotalQty}