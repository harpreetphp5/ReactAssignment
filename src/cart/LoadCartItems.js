import React, { useContext } from 'react';
import { ProductsContext } from '../../src/helpers/ProductsContext';

export const updateCartQty = (identifier, prod, dispatch) => {

    if (prod.qty < 2 && identifier == -1) {
        dispatch({ type: 'DELETE_CART', payload: prod.title });
        return 'deleted last item';
    }
    dispatch({ type: 'UPDATE_CART', payload: { title: prod.title, items: prod, qty: identifier } });
    return 'updated quantity';
}

export const LoadData = (items, dispatch) => {
    return (
        items.map(
            p => (
                <>
                    <hr />
                    <div className="cartWrap">
                        <li className="productTab cartPage">
                            <div>
                                <img src={'/media/' + p.items.image} />
                            </div>
                            <div>
                                <p className="brand">{p.items.brand}</p>
                                <h3 className="title">{p.title}</h3>
                                <p className="color">COLOR: {p.items.color}</p>
                            </div>
                        </li>
                        <li className="qtyTab">
                            <div className="qtyBox">
                                <input className="qtyType" type="text" value={p.qty} />
                            </div>
                            <div className="qtyBtns">
                                <button onClick={() => updateCartQty(+1, p, dispatch)}>+</button>
                                <button onClick={() => updateCartQty(-1, p, dispatch)} className="second">-</button>
                            </div>
                        </li>
                        <li className="totalTab">
                            ${(p.qty * p.items.price).toFixed(2)}
                        </li>
                        <li className="actionTab">
                            <button className="crossDelSimple" onClick={() => dispatch({ type: 'DELETE_CART', payload: p.title })}>X</button>
                        </li>
                    </div>

                </>
            ))
    )
}

export const LoadCartItems = () => {
    const [allProducts, cartItems, dispatch] = useContext(ProductsContext);
    
    return (
        LoadData(cartItems, dispatch)
        )
}
