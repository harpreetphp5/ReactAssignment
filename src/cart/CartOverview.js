import React, { useContext } from 'react';
import { ProductsContext } from '../../src/helpers/ProductsContext';
import { getTotalPrice } from '../helpers/cartTotal';
import { Link } from 'react-router-dom';

const CartOverview = () => {
    const [allProducts, cartItems, dispatch] = useContext(ProductsContext);
    return (
        <li>
            <hr />
            <div className="cartOverview">
                <li className="productTab"></li>
                <li className="qtyTab"><span>CART OVERVIEW</span><span></span><br /></li>
                <li className=" priceSubtotal"><span>SUBTOTAL</span><span>${getTotalPrice(cartItems).toFixed(2)}</span></li>
                <li className=" priceTotal"> <span>TOTAL</span><span>${getTotalPrice(cartItems).toFixed(2)}</span></li>
            </div>
            <hr className="clear" />
            <div className="cartActions">
                <span><Link to="/" className="continueShopping">CONTINUE SHOPPING</Link></span>
                <Link to="/cart" className="actionButtonCart">CHECKOUT</Link>
            </div>
        </li>
    )
}

export default CartOverview;