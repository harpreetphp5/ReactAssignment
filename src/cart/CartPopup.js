import React, { useContext } from 'react';
import { Link, Route } from 'react-router-dom';
import { getTotalPrice, getTotalQty } from '../helpers/cartTotal';
import { ProductsContext } from '../../src/helpers/ProductsContext';

const CartPopup = () => {
    const [allProducts, cartItems, dispatch] = useContext(ProductsContext);
    return(
    <header className="Cart-header">
        <ul to="/cart" className="CartList">
            <li><Link className="myCartLink" to="/cart">MY CART ({getTotalQty(cartItems)})</Link><ul>
                {
                    cartItems.map(
                        p => (
                            <li className="cartLayout">
                                <div>
                                    <img src={'/media/' + p.items.image} />
                                </div>
                                <div>
                                    <h3 className="title">
                                        {p.title}
                                        <span className="qty">
                                            x{p.qty}
                                        </span>

                                    </h3>


                                    <p className="brand">{p.items.brand}</p>
                                    <p className="price cartColor">${(p.items.price).toFixed(2)}</p>
                                    <span>
                                        <button className="crossDel" onClick={() => dispatch({ type: 'DELETE_CART', payload: p.title })}>X</button>
                                    </span>
                                </div>
                            </li>
                        )
                    )
                }
                <li>
                    <hr />
                    <div className="cartTotal">
                        <span>TOTAL</span><span>${(getTotalPrice(cartItems)).toFixed(2)}</span>
                    </div>
                    <div className="cartActions">
                        <Link to="/cart" className="actionButtonCartWhite">CHECKOUT</Link>
                        <Link to="/cart" className="actionButtonCart">VIEW CART</Link>
                    </div>
                </li>
            </ul></li>

        </ul>

        </header>
    )
}

export default CartPopup;