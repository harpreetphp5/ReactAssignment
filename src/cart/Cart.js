import React from 'react';
import { LoadCartItems } from './LoadCartItems';
import CartOverview from './CartOverview';
import CartHeader from './CartHeader';
import './Cart.css';

const Cart = () => {

    return (
        <div className="Cart">
            <ul className="cartHeader">
                <CartHeader />
            </ul>
            <ul>
                <LoadCartItems />
                <CartOverview />
            </ul>
        </div>
    )
};

export default Cart;
