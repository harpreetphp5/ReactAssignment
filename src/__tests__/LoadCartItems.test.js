import React, { useContext } from 'react';
import ReactDOM from 'react-dom';
import { LoadCartItems, LoadData } from '../cart/LoadCartItems'
import { reducer } from '../helpers/cartReducer';

it('renders without crashing', () => {
    const mockCartItems = [
        {
            title: "Hand Painted Blue Flat Dish",
            items: {
                "title": "Hand Painted Blue Flat Dish",
                "brand": "Kiriko",
                "price": 28,
                "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam at purus pulvinar, placerat turpis ac, interdum metus. In eget massa sed enim hendrerit auctor a eget arcu. Curabitur ac pharetra nisl, sit amet mattis dolor.",
                "image": "hand-painted-blue-flat-dish.jpg",
                "color": "Blue" },
            qty: 1
        }
    ]
    expect(LoadData(mockCartItems)).toMatchSnapshot();
});



describe('reducer testing', () => {

    it('should return the initial state', () => {
        expect(reducer('', {})).toEqual('');
    });

    it('should Add to cart', () => {
        const mockCartItems2 =
        {
            title: "Hand Painted Blue Flat Dish",
            items: {
                "title": "Hand Painted Blue Flat Dish",
                "brand": "Kiriko",
                "price": 28,
                "description": "test data",
                "image": "hand-painted-blue-flat-dish.jpg",
                "color": "Blue"
            },
            qty: 1
        };
        expect(reducer([], { type: "ADD_CART", payload: mockCartItems2 })).toEqual([mockCartItems2]);

    });

    it('should set update quantity to 3', () => {
        const mockCartItems =
        {
            title: "Hand Painted Blue Flat Dish",
            items: {
                "title": "Hand Painted Blue Flat Dish",
                "brand": "Kiriko",
                "price": 28,
                "description": "test data",
                "image": "hand-painted-blue-flat-dish.jpg",
                "color": "Blue"
            },
            qty: 1
        };

        const state = reducer([mockCartItems], { type: "UPDATE_CART", payload: { title: "Hand Painted Blue Flat Dish", qty: 2 } });

        expect(state.map((p) => p.qty ? p.qty : p)).toEqual([3]);
    });

    it('expect state to increase in length by 1 when added to cart', () => {
        const mockCartItems =
        {
            title: "Hand Painted Blue Flat Dish",
            items: {
                "title": "Hand Painted Blue Flat Dish",
                "brand": "Kiriko",
                "price": 28,
                "description": "test data",
                "image": "hand-painted-blue-flat-dish.jpg",
                "color": "Blue"
            },
            qty: 1
        };

        const newOutput = { "items": { "brand": "Kiriko", "color": "Blue", "description": "test data", "image": "hand-painted-blue-flat-dish.jpg", "price": 28, "title": "New Title" }, "qty": 3, "title": "Hand Painted Blue Flat Dish" };

        const state = reducer([mockCartItems], { type: "ADD_CART", payload: newOutput } );

        expect(state.length).toEqual(2);
    });

    it('expect state to decrease in length by 1 when delete from cart', () => {
        const mockCartItems =
        {
            title: "Hand Painted Blue Flat Dish",
            items: {
                "title": "Hand Painted Blue Flat Dish",
                "brand": "Kiriko",
                "price": 28,
                "description": "test data",
                "image": "hand-painted-blue-flat-dish.jpg",
                "color": "Blue"
            },
            qty: 1
        };


        const state = reducer([mockCartItems], { type: "DELETE_CART", payload: 'Hand Painted Blue Flat Dish' });

        expect(state).toEqual([]);
    });
});



