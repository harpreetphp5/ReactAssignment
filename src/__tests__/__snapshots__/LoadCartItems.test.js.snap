// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Array [
  <React.Fragment>
    <hr />
    <div
      className="cartWrap"
    >
      <li
        className="productTab cartPage"
      >
        <div>
          <img
            src="/media/hand-painted-blue-flat-dish.jpg"
          />
        </div>
        <div>
          <p
            className="brand"
          >
            Kiriko
          </p>
          <h3
            className="title"
          >
            Hand Painted Blue Flat Dish
          </h3>
          <p
            className="color"
          >
            COLOR: 
            Blue
          </p>
        </div>
      </li>
      <li
        className="qtyTab"
      >
        <div
          className="qtyBox"
        >
          <input
            className="qtyType"
            type="text"
            value={1}
          />
        </div>
        <div
          className="qtyBtns"
        >
          <button
            onClick={[Function]}
          >
            +
          </button>
          <button
            className="second"
            onClick={[Function]}
          >
            -
          </button>
        </div>
      </li>
      <li
        className="totalTab"
      >
        $
        28.00
      </li>
      <li
        className="actionTab"
      >
        <button
          className="crossDelSimple"
          onClick={[Function]}
        >
          X
        </button>
      </li>
    </div>
  </React.Fragment>,
]
`;
